### DISCUSSION TITLE
Pixel to pixel misorientation calculation inside an arbitrary rectangular area

### DISCUSSION THREAD

**Discussion by Mustafa-Rifat** [mtex-toolbox/mtex#693] [Ask Anything]
Hello,

In my code, I am able to calculate misorientation along a line (defined by 2 points selected by me). However, 
1. I want to select 4 points now which will form a rectangle
2. I want to get the misorientation in degrees from pixel to pixel inside that rectangle
3. I want to make a surface plot of the misorientations, there will be 2 surface plots one showing misorientation along the length of the rectangle; another showing misorientation along the width of the rectangle 

The code I use currently after plotting the ebsd goes like below. It plots misorientation along a line:
```matlab
p1=ginput(1);
p2=ginput(1);
g1=findByLocation(ebsd,p1);
g2=findByLocation(ebsd,p2);
angle(ebsd{g1}.orientations,ebsd{g2}.orientations)/degree
step=0.1;
distance=sqrt((ebsd{g1}.x-ebsd{g2}.x)^2+(ebsd{g1}.y-ebsd{g2}.y)^2);
slope=(ebsd{g1}.y-ebsd{g2}.y)/(ebsd{g1}.x-ebsd{g2}.x);
theta=atan(slope);
if ebsd{g2}.x<ebsd{g1}.x
    theta=theta+pi;
end
misorientation_profile=[];
orientation_=[];
for d=0:step:distance
    x=ebsd{g1}.x+cos(theta)*d;
    y=ebsd{g1}.y+sin(theta)*d;
    p_intermediate=[x y];
    g_intermediate=findByLocation(ebsd,p_intermediate);
    orientation_=[orientation_; ebsd{g_intermediate}.orientations];
    misorientation_profile=[misorientation_profile;[d angle(ebsd{g1}.orientations,ebsd{g_intermediate}.orientations)/degree]];
    plot(x,y,'.w','markersize',5)
end
%saveas(gca,'point_map.tif','tif');
figure
ax=gca;
plot(misorientation_profile(:,1),misorientation_profile(:,2),'black','lineWidth',3);hold on;plot(misorientation_profile(:,1),misorientation_profile(:,2),'or','MarkerSize',5);
```

**Comment by kilir** [comment]
Hi,
with respect to your misorientation profile, you can have a look at `EBSD/spatialProfile`. 
With respect to your pixel to pixel misorientation, please explain how the pixel to pixel misorientation shall relate to the rectangle?
I am also unsure what you mean by the surface plots of misorientations. Could you elaborate?

Cheers,
Rüdiger

**Reply by Mustafa-Rifat** [reply]
Hello Rüdiger,

Thanks a lot for your reply.  See the attached image below.
![138649982_1685323404963020_6035439591210910565_n](https://user-images.githubusercontent.com/60717121/104361223-2da30d00-54e0-11eb-9474-c1bacaed12ad.jpg)
There are 2 grains. I want to select 4 points which will create a box (rectangle). I want to consider only the pixels inside this box. Once I have that, I want to make surface plot 1 and 2 as sketched. 
Surface plot 1: It will plot the misorientation of all the pixels situated above the bottom part of the box. In the bottom there are 5 pixels (1,a,b,c,3). In the top there are 5 pixels (2,d,e,f,4). In my surface plot I want to show the misorientation of all the pixels from 1 to 2 with respect to pixel 1; misorientation of all the pixels from a to d with respect to d and so on.
Surface plot 2: Similarly, It will plot the misorientation of all the pixels situated left to the the rightmost part of the box. i,e, misorientation of all the pixels from 1 to 3 with respect to 1; misorientation of all the pixels from g to l with respect to g and so on.

**Reply by kilir** [reply]
Morning, 
I think I understand-basically a stack of misoientation profiles, right? Are the box edges always parallel to columns/rows?
Cheers, Rüdiger

**Reply by Mustafa-Rifat** [reply]
Hi Rüdiger,

Yes you are right. I solved the issue. I plotted one surface plot though. I realized that would be enough! Here is the code that I used:
```
%select a point at first and create the second point 3 unit above it
p1=ginput(1);`
p2=[p1(1) p1(2)+3];
g1=findByLocation(ebsd,p1);
g2=findByLocation(ebsd,p2);
step=0.1;
%this will be the height of the box
distance1=sqrt((ebsd{g1}.x-ebsd{g2}.x)^2+(ebsd{g1}.y-ebsd{g2}.y)^2);
%width of the box is 2
distance2=2;
slope=(ebsd{g1}.y-ebsd{g2}.y)/(ebsd{g1}.x-ebsd{g2}.x);
%measure the inclination angle
theta=atan(slope);
if ebsd{g2}.x<ebsd{g1}.x
    theta=theta+pi;
end
misorientation_profile=[];
orientation_=[];
%the loop is the heart of the code. It calculates the misorientation of all the points inside the box that I have just created. It also lists the orientation of the points inside the box.
for d2=0:step:distance2
    for d1=0:step:distance1
    x=ebsd{findByLocation(ebsd,[p1(1)+d2 p1(2)])}.x+cos(theta)*d1;
    y=ebsd{findByLocation(ebsd,[p1(1)+d2 p1(2)])}.y+sin(theta)*d1;
    p_intermediate=[x y];
    g_intermediate=findByLocation(ebsd,p_intermediate);
    orientation_=[orientation_; ebsd{g_intermediate}.orientations];
    misorientation_profile=[misorientation_profile;[d1 angle(ebsd{findByLocation(ebsd,[p1(1)+d2 p1(2)])}.orientations,ebsd{g_intermediate}.orientations)/degree]];
%use the next line instead of the previous line if you want to plot the misorientation with respect to the mean orientation of the parent grain (i.e. grain where the white box is situated)    
%misorientation_profile=[misorientation_profile;[d1 angle(ebsd{g_intermediate}.orientations,grains (p1(1),p1(2)).meanOrientation)/degree]];
    plot(x,y,'.w','markersize',5)
    end
end 
%surface plot
[X,Y] = meshgrid(0:step:distance2,0:step:distance1);
mis_prof=misorientation_profile(:,2);
mis_prof=reshape(mis_prof,size(X));
figure
surface(X,Y,mis_prof,'edgecolor','none');axis equal
colorbar
xlabel('Region Width (\mu m)');
ylabel('Region Height (\mu m)')
%title('Misorientation (deg.): Mean=' num2str(mean(mean(mis_prof))),',Stdev=' num2str(std2(mis_prof)))
title(['Misorientation (deg.): mean=' num2str(mean(mean(mis_prof))) ', St.dev.=' num2str(std2(mis_prof))]);
set(gca,'FontSize',22,'FontWeight','Bold','FontName','Calibri','FontAngle','italic');
```
